docs: Reliable Channel API spec (#104)
waku-org/specs: standards/application/reliable-channel-api.md

    * [Incoming message processing](#incoming-message-processing)
    * [Retransmission](#retransmission)
    * [Rate limiting](#rate-limiting)
    * [Encryption](#encryption)
  * [Security/Privacy Considerations](#securityprivacy-considerations)
  * [Copyright](#copyright)
<!-- TOC -->

## Abstract

This document specifies the **Reliable Channel API**,
an application-level interface that sits between the application layer and the [MESSAGING-API](/standards/application/messaging-api.md) plus [P2P-RELIABILITY](/standards/application/p2p-reliability.md), i.e., `application` <-> **reliable-channel-api** <-> `messaging-api/p2p-reliability`.

It bundles segmentation, end-to-end reliability via [Scalable Data Sync (SDS)](https://lip.logos.co/ift-ts/raw/sds.html), rate limit management, and a pluggable encryption hook
into a single interface for sending and receiving messages reliably.

## Motivation

The [MESSAGING-API](/standards/application/messaging-api.md) provides peer-to-peer reliability via [P2P-RELIABILITY](/standards/application/p2p-reliability.md),
but does not provide high end-to-end delivery guarantees from sender to recipient.

This API addresses that gap by introducing:

> - **[SEGMENTATION-API](/standards/application/segmentation-api.md)** to handle large messages exceeding network size limits.

Reviewer: As in https://github.com/logos-messaging/specs/pull/91/changes
```suggestion
- **[SEGMENTATION-API](/standards/application/segmentation.md)** to handle large messages exceeding network size limits.
```

```
┌────────────────────────────────────────────────────────────┐
│                   Application Layer                        │
└───────────────────────────┬────────────────────────────────┘
                            │
┌───────────────────────────▼────────────────────────────────┐
│                 Reliable Channel API                       │
│  ┌──────────────┐ ┌─────┐ ┌───────────────┐ ┌──────────┐   │
│  │ Segmentation │ │ SDS │ │ Rate Limit Mgr│ │Encryption│   │
│  │              │ │     │ │               │ │   Hook   │   │
│  └──────────────┘ └─────┘ └───────────────┘ └──────────┘   │
└───────────────────────────┬────────────────────────────────┘
                            │
┌───────────────────────────▼────────────────────────────────┐
│                    Messaging API                           │
│      (P2P Reliability, Relay, Filter, Lightpush, Store)    │
└────────────────────────────────────────────────────────────┘
```

### IDL

A custom Interface Definition Language (IDL) in YAML is used, consistent with [MESSAGING-API](/standards/application/messaging-api.md).

## The Reliable Channel API

> This API considers the types defined by [MESSAGING-API](/standards/application/messaging-api.md) plus the following (including `SdsConfig` and `RateLimitConfig`).

Reviewer: ```suggestion
This API considers the types defined by [MESSAGING-API](/standards/application/messaging-api.md) plus the following.
```

### Channel lifecycle

This point assumes that a WakuNode instance is created beforehand. See `createNode` function
in [MESSAGING-API](/standards/application/messaging-api.md).

```yaml
functions:

  createReliableChannel:
    description: "Opens a reliable channel over the given content topic. Sets up the required SDS state,
    segmentation, and encryption, and subscribes to `contentTopic` via the underlying

>     [MESSAGING-API](/standards/application/messaging-api.md)."

Reviewer: ```suggestion
    description: "Creates a reliable channel over the given content topic. Sets up the required SDS state,
    segmentation, and encryption, and subscribes to `contentTopic`."
```

    parameters:
      - name: channel
        type: ReliableChannel
        description: "The channel handle returned by `createReliableChannel`."
    returns:
      type: result<void, error>
```

### Channel usage

```yaml
functions:
  send:
    description: "Send a message through a reliable channel. The message is always segmented,
    SDS-tracked, rate-limited, and encrypted (if configured)."
    parameters:
      - name: channel
        type: ReliableChannel
        description: "The channel handle returned by `createReliableChannel`."
      - name: message
        type: array<byte>
        description: "The raw message payload to send."
    returns:
      type: result<ReliableSendId, error>
      description: "Returns a `ReliableSendId` that callers can use to correlate subsequent `MessageSentEvent` or `MessageSendErrorEvent` events."
```

Incoming events are emitted on `channel.messageEvents` as defined by `MessageEvents`.

### Node configuration

This spec extends `NodeConfig`, needed to create a node, which is 
defined in [MESSAGING-API](/standards/application/messaging-api.md), 
with `sds_config` and `rate_limit_config` fields, whose types are defined 

> in [Type definitions](#type-definitions).

Reviewer: ```suggestion
with `sds_config` and `rate_limit_config` fields.
```

```yaml
NodeConfig:  # Extends NodeConfig defined in MESSAGING-API
  fields:
    sds_config:
      type: SdsConfig
      description: "SDS configuration. See SdsConfig defined in this spec."
    rate_limit_config:
      type: RateLimitConfig

>       description: "Rate limiting configuration, including RLN-specific attributes. See RateLimitConfig defined in this spec."

Reviewer: Just realized that it probably shouldn't be configurable at all, but rather always taken from node configuration itself, e.g. `--rln-relay-epoch-sec`.

Author: Yes makes sense. Maybe hard-coded per version?

Reviewer: > Maybe hard-coded per version?

What do you mean? 🤔

Author: I mean to keep having the RLN params hard-coded. Disregard that suggestion. Is better having good defaults and allow changing the config.

Author: Addressed in 0c512d935c077741ca29f3a34021c86e5be2c081

```

### Type definitions

```yaml
types:

  ReliableChannel:
    type: object
    description: "A handle representing an open reliable channel.
    Returned by `createReliableChannel` and used to send messages and receive events.
    Internal state (SDS, segmentation, encryption) is managed by the implementation."
    fields:
      messageEvents:
        type: MessageEvents

>         description: "Event emitter for reliable message events scoped to this channel."

Reviewer: So to subscribe I'll need to do smth like `channel.messageEvents.onMessageReceived(...)`?
Can it be simplified to `channel.onMessageReceived(...)`?

Author: Great point! Addressed in bd6abfbb292044da04d1b774f62db36fc4c7db8a

Reviewer: Hmmm. Sorry I didn't mean to particularly introduce callbacks here.
Callbacks are implementation-specific. And specf should be agnostic of it.

IDL in Messaging API defines events and, we should leverage it, no?

There's 2 ways of doing it:
1. Let `object` have `events: ` section, and just define all events directly there.
    ```yaml
    ReliableChannel:
        type: object
        events:
          "reliable:message:received":
            type: MessageReceivedEvent
    ```
2. Change type of `ReliableChannel` to `event_emitter`. 

I prefer (1).

Author: Addressed in 493b909f0282165c34a27b648997eeec2bc4b645

      rateLimitConfig:
        type: RateLimitConfig
        default: DefaultRateLimitConfig

>         description: "Configuration for rate limit management."

Reviewer: Rate Limiting will happen per-node, not per-channel, so it shouldn't be here.
```suggestion
```

  MessageEvents:
    type: event_emitter
    description: "Event source for reliable message events on a channel"
    events:
      "reliable:message:received":
        type: MessageReceivedEvent
      "reliable:message:sent":
        type: MessageSentEvent
      "reliable:message:send-error":
        type: MessageSendErrorEvent

  MessageReceivedEvent:
    type: object
    description: "Event emitted when a complete message has been received and reassembled."
    fields:
      message:
        type: array<byte>
        description: "The reassembled message payload."

>   MessageSentEvent:

Reviewer: An acknowledged message is means it's not just "sent", but already delivered.
```suggestion
  MessageDeliveredEvent:
```

We could still have a `MessageSentEvent`, if we're planning to call Messaging API asynchronously. But this would mean that we have successfully relayed/lightpushed all chunks of the message.

Author: Great catch thanks! Addressed in 408af18b8d0ef8e05998bf7e41e3bac59c31038a
EDITED: we need to differentiate between `MessageSentEvent` ( delivered to the network ) vs `MessageDeliveredEvent` ( delivered to the target receiver )

    type: object
    description: "Event emitted when all chunks of a message have been acknowledged by the network."
    fields:
      requestId:
        type: ReliableSendId
        description: "The identifier of the `send` operation whose chunks have all been acknowledged."

>   MessageSendErrorEvent:

Reviewer: Similar to https://github.com/logos-messaging/specs/pull/104/changes#r3150230317
```suggestion
  MessageDeliveryErrorEvent:
```

Author: In this particular case, I think it is not needed for now. The error event itself will have enough detail to know why it happened.

Reviewer: But what does this event mean then?

> Event emitted when a message send operation fails after exhausting retransmission attempts.

Is it send operation failed? Meaning e.g. we have not peers.
Or is it rentranmissions attempte exhausted? Meaning we didn't get SDS acknowledge even after many retransmissions?

Author: The error can contain any of the options that you mentioned.
Extended the error description in the following: 33f89ccbd8d94f0a085f242b9105b2f6f85cb263

    type: object
    description: "Event emitted when a message send operation fails after exhausting retransmission attempts."
    fields:
      requestId:
        type: ReliableSendId
        description: "The identifier of the `send` operation that failed after exhausting retransmission attempts."
      error:
        type: string
        description: "Error message describing what went wrong"

>   ReliableSendId:

Reviewer: I think at this point it can be simply a "Message ID"
```suggestion
  MessageId:
```

Author: We may need to have certain analogy with `messaging-api`. In there, we use `RequestId`. Maybe we can use a similar term, f.e., `ReliableRequestId`, considering that one `ReliableRequestId` will be composed of several `RequestId`s.

Reviewer: All the options starting with `Reliable...` doesn't make sense to me. I understand the intention to use api naming, but requests and messages are not "reliable". 

Anything wrong calling it a "message id"? 
To me this makes most sense, as it's the most high-level API that we'll have. 

Also, it could actually be just `string` without custom type 😄

Author: As per our 1:1 chat, I'm suggesting the use of `RequestId` in the following: dba72eb697679d4e44abe97d2a6abbf2aeaf5de4

Reviewer: I still think that `MessageId` is better, but this works 👍

    fields:
      encrypt:
        type: function
        description: "Encrypts a byte payload. Returns the encrypted payload."
        parameters:
          - name: content
            type: array<byte>
        returns:
          type: result<array<byte>, error>
      decrypt:
        type: function
        description: "Decrypts a byte payload. Returns the decrypted payload."
        parameters:
          - name: payload
            type: array<byte>
        returns:
          type: result<array<byte>, error>

  Persistence:
    type: object
    description: "Interface for a pluggable SDS persistence backend.
    Implementations MUST provide all functions required to save and retrieve SDS state per channel. Implementations MUST also provide the persistence method of interest, e.g., SQLite, custom encrypted storage, etc.
    Refer to the [SDS spec](https://lip.logos.co/ift-ts/raw/sds.html) for the full definition of what state must be persisted."
```

## Components

### Segmentation

See [SEGMENTATION-API](./segmentation-api.md).

### Scalable Data Sync (SDS)

[SDS](https://lip.logos.co/ift-ts/raw/sds.html) provides end-to-end delivery guarantees using causal history tracking.

> - Each sent chunk is registered in an outgoing buffer.

Reviewer: Let's stick to the work "segment" instead of chunk, according to SEGMENTATION spec (https://github.com/logos-messaging/specs/pull/91.

Author: Addressed in f8ab3562310c35d16561f09754b14878c6187683

- The sender removes acknowledged chunks from the outgoing buffer.
- Unacknowledged chunks are retransmitted after `acknowledgementTimeoutMs`.
- SDS state MUST be persisted using the `persistence` backend configured in `SdsConfig`.

### Rate Limit Manager

The Rate Limit Manager ensures compliance with [RLN](https://lip.logos.co/messaging/standards/core/17/rln-relay.html) rate constraints.

- It tracks how many messages have been sent in the current epoch (only the first chunk of each message counts toward the rate limit; subsequent chunks are exempt).
- When the limit is approached, chunk dispatch MUST be delayed to the next epoch.
- The epoch size MUST match the `epochSizeMs` configured in `RateLimitConfig`.

### Encryption Hook

The Encryption Hook provides a pluggable interface for upper layers to inject encryption.

- The hook is optional; when not provided, messages are sent unencrypted.
- Encryption is applied per chunk, after segmentation and SDS registration.
- Decryption is applied per chunk, before SDS delivery.
- The `Encryption` interface MUST be implemented by the caller.
- The Reliable Channel API MUST NOT impose any specific encryption scheme.

> ## Procedures

Reviewer: Wait, didn't we move this block **before** `## The Reliable Channel API`? 🤔 
You commented about it here: https://github.com/logos-messaging/specs/pull/104#discussion_r3136983513

Author: Addressed in 3c1125bddb751de5d5b92565fb0acaeff2273796

### State management

Each `ReliableChannel` MUST maintain internal state for [SDS](https://lip.logos.co/ift-ts/raw/sds.html), keyed by `channelId`. This state MAY be persisted within the SDS implementation layer and includes:
- A committed message log recording sent messages and their acknowledgement status.
- An outgoing buffer of unacknowledged message chunks pending confirmation.
- An incoming buffer for partially received segmented messages awaiting full reassembly.

> The state MUST be persisted using the `persistence` backend specified in `SdsConfig`.

Reviewer: I'm not sure what you're trying to say here, but this whole section doesn't make sense to me.

No need to specify SDS internal behaviour, there is SDS specification for it. What we should only do is describe that we should wrap/unwrap SDS messages. And you do it below in "...  message processing" sections.

Author: Addressed in 198ad99bed87c2c2dbb02675291ecbc13ce0ca7e

### Outgoing message processing

When `send` is called, the implementation MUST process `message` in the following order:

> 1. **Segment**: Split the payload into chunks as defined in [SEGMENTATION-API](./segmentation-api.md).

Reviewer: Note that Segmentation here should keep enough space for SDS headers to be added for each chunk.
More details here: https://github.com/logos-messaging/specs/pull/91/changes#r2474253119

Author: Addressed in 0a57672eade10df38c646193f3b5dc75abaee757

2. **Apply [SDS](https://lip.logos.co/ift-ts/raw/sds.html)**: Register each chunk with the SDS layer to track acknowledgements and enable retransmission.
3. **Encrypt**: If an `Encryption` implementation is provided, encrypt each chunk before transmission.
4. **Rate Limit**: If `RateLimitConfig.enabled` is `true`, delay dispatch as needed to comply with [RLN](https://lip.logos.co/messaging/standards/core/17/rln-relay.html) epoch constraints.
5. **Dispatch**: Send each chunk via the underlying [MESSAGING-API](/standards/application/messaging-api.md).

### Incoming message processing

When a chunk is received from the network, the implementation MUST process it in the following order:

1. **Decrypt**: If an `Encryption` implementation is provided, decrypt the chunk.
2. **Apply [SDS](https://lip.logos.co/ift-ts/raw/sds.html)**: Deliver the chunk to the SDS layer, which emits acknowledgements and detects gaps.

> 3. **Reassemble**: Once all chunks for a message have been received, reassemble and emit a `reliable:message:received` event.

Reviewer: Chunks should be persisted on disk until we assemble the whole message back. Otherwise a message will be lost if closed before receiving all chunks.

Author: Okay but this is part of segmentation api, isn't it :) ?

Reviewer: 1. Segmentation is not an API, it's a protocol. And I think its implementation won't know anything about persistence.
2. Even if Segmentation implementation used persistence, it doesn't know anything about "message being delivered".

So I think it should be fully done on Reliable Channel API side, no?

Author: Yes, I agree. The reassemble and persistence will be handled by Reliable Channel API.

Reviewer: Pardon, apparently segmentation implementation [actually has](https://github.com/status-im/status-go/blob/0fcf4ceb8e0df4f927460f40f980256cf5518e84/pkg/messaging/layers/segmentation/persistence.go#L7-L14) `persistence` interface: 
```go
type Persistence interface {
	IsMessageAlreadyCompleted(hash []byte) (bool, error)
	SaveMessageSegment(segment *Message, sigPubKey *ecdsa.PublicKey, timestamp int64) error
	GetMessageSegments(hash []byte, sigPubKey *ecdsa.PublicKey) ([]*Message, error)
	CompleteMessageSegments(hash []byte, sigPubKey *ecdsa.PublicKey, timestamp int64) error
	RemoveMessageSegmentsOlderThan(timestamp int64) error
	RemoveMessageSegmentsCompletedOlderThan(timestamp int64) error
}
```

I'm not sure if it's the same in Nim implementation. If not, we should add it.
So Reliable Channel API need to fullfil this interface, but no need to store segments ourselves.

> 

Reviewer: Would be nice to add a section regarding reacting to these SDS signal `MissingDependencies`. We should automatically fetch missing messages from a store node.

Reviewer: Side-note: there's also a `onRetrievalHint` signal in `nim-sds`, which is never triggered. It should be removed.
Retrieval hint is passed together with message

Author: Great points! Addressed in 6590306abb9d86ea2001dddac23319d21ef026b0

> 

Reviewer: Also, would be nice to add a section for actions at the moment of creating a `ReliableChannel`.

1. Initialize SDS:~~SDS should read its state for this channel from `persistence`.~~ <-- this is implementation detail, no need to mention it. But we should make sure to implement such functionality in `nim-sds`.
2. Initialize rate limit manager: application will always be started at a random moment of epoch, and we should know our remaining quota.

Author: Maybe we can leave this part a little bit opened for now. I also believe this belongs more to MessagingAPI and the `initialize` should happen at node's creation, isn't it?

Reviewer: Messaging API can't initialize SDS channels, right?
We can leave it opened for now, but important not to forget about it.

Author: Yes thanks!
I created these for future reference:
- https://github.com/logos-messaging/nim-sds/issues/65
- https://github.com/logos-messaging/logos-delivery/issues/3838

### Retransmission

A certain Reliable Channel MUST retransmit unacknowledged chunks after `acknowledgementTimeoutMs`,
according to the [SDS](https://lip.logos.co/ift-ts/raw/sds.html) channel state,
up to `maxRetransmissions` times.

> If a chunk is not acknowledged after all retransmission attempts, a `reliable:message:send-error` event MUST be emitted.

Reviewer: Same as https://github.com/logos-messaging/specs/pull/104/changes#r3150169122.
This seem to be done by `nim-sds` in [sds.nim#L293-L338](https://github.com/logos-messaging/nim-sds/blob/13559198b06dfb7d3bf8018c5f2f1e9ca9e54999/sds.nim#L293-L338)

Author: Right, no need to explain the same twice. Addressed in 330e58ce4bf61592267e172df362ffefc5e14848

---
title: RELIABLE-CHANNEL-API
name: Reliable Channel API definition
category: Standards Track
status: raw
tags: [reliability, application, api, sds, segmentation]

> editor: Logos Messaging Team

Reviewer 2: [Dust] While not strictly forbidden by 1/COSS. My understanding was multiple editors adds ambiguity to the process. "Who is responsible for changes?", "what if team members disagree?" Choosing a single person responsible for managing the lifecycle of the document, avoids the "diffusion of responsibility"

Author: ah okay, addressed now

Reviewer 2: > - Igor Sirotin <sirotin@status.im>
> - Jazz Turner-Baggs <jazz@status.im>
> - Ivan Folgueira Bande <ivansete@status.im>

This is more the specs team area of expertise.

However: I think you want a single editor, with multiple Contributors. The editor is just the person who is responsible for responding to updates. This can be sorted out later when it gets pushed to Logos-Lips

> editor: Logos Messaging Team

Reviewer 2: [Sand] According to 1/COSS this field should contain both name and email address.

Author: Done, thanks

It bundles segmentation, end-to-end reliability via [Scalable Data Sync (SDS)](https://lip.logos.co/ift-ts/raw/sds.html), rate limit management, and a pluggable encryption hook
into a single interface for sending and receiving messages reliably.

## Motivation

The [MESSAGING-API](/standards/application/messaging-api.md) provides peer-to-peer reliability via [P2P-RELIABILITY](/standards/application/p2p-reliability.md),
but does not provide high end-to-end delivery guarantees from sender to recipient.

This API addresses that gap by introducing:
- **[SEGMENTATION](/standards/application/segmentation.md)** to handle large messages exceeding network size limits.
- **[SDS](https://lip.logos.co/ift-ts/raw/sds.html)** to provide causal-history-based end-to-end acknowledgement and retransmission.
- **Rate Limit Manager** to comply with [RLN](https://lip.logos.co/messaging/standards/core/17/rln-relay.html) constraints when sending segmented messages.

> - **Encryption Hook** to allow upper layers to provide a pluggable encryption mechanism. The main motivation for encryption is to protect the payload. 

Reviewer 2: >  protect the payload. 

[Dust] from what? How can a reviewer verify its been adequately protected?

Reviewer 2: ```suggestion
- **Encryption Hook** to allow upper layers to provide a pluggable encryption mechanism. This enables applications to provide Confidentiality and Integrity if desired. 
```

Author: Addressed in f39c7f0d0bef5a05bae276439dea471790df3ab1

```
┌────────────────────────────────────────────────────────────┐
│                   Application Layer                        │
└───────────────────────────┬────────────────────────────────┘
                            │
┌───────────────────────────▼────────────────────────────────┐
│                 Reliable Channel API                       │
│  ┌──────────────┐ ┌─────┐ ┌───────────────┐ ┌──────────┐   │
│  │ Segmentation │ │ SDS │ │ Rate Limit Mgr│ │Encryption│   │
│  │              │ │     │ │               │ │   Hook   │   │
│  └──────────────┘ └─────┘ └───────────────┘ └──────────┘   │
└───────────────────────────┬────────────────────────────────┘
                            │
┌───────────────────────────▼────────────────────────────────┐
│                    Messaging API                           │
│      (P2P Reliability, Relay, Filter, Lightpush, Store)    │
└────────────────────────────────────────────────────────────┘
```

### IDL

A custom Interface Definition Language (IDL) in YAML is used, consistent with [MESSAGING-API](/standards/application/messaging-api.md).

## Components

### Segmentation

> See [SEGMENTATION](./segmentation.md).

Reviewer 2: [Dust] A brief explanation of what the component is would be helpful.

Author: Addressed in a6fd069eebbf641d3b9073dad9b83c023d39b3cb

> See [SEGMENTATION](./segmentation.md).

Reviewer 2: [Pebble] This spec does not define the configuration parameters that are needed for the segmentation spec.

While the PR has not been merged, this PR can define 
- `segmentSize`  which is a required parameter.

Author: Addressed in a6fd069eebbf641d3b9073dad9b83c023d39b3cb

Reviewer: Configuration parameters are described in `SegmentationConfig`. Though indeed there's no explanation that `SegmentationConfig` matches to certain segmentation protocol properties

Reviewer 2: Theres improvements to be made, however this latest change provides enough information to start linking concepts together.

### Scalable Data Sync (SDS)

[SDS](https://lip.logos.co/ift-ts/raw/sds.html) provides end-to-end delivery guarantees using causal history tracking.

- Each new segment to be sent, requires the following data:

>   - `MessageId`: a 40-character random hex string (e.g. `a23f89b4df7e1ba5a709352da3a247a4fd609f4c`), generated by the Reliable Channel API.

Reviewer 2: [Dust] Using a random hex string doesn't violate the SDS protocol definition. But it is somewhat of a interesting decision. I would document why you are choosing this over a more common `Hash(DomainSep + MessageBytes)` 

Specifically I would address how there is no concern about multiple sends generating different copies of SdsMessages.

Author: Addressed in 491b2e36d13f29bf2fb93b6cbb2534ebe847ed19 ( cc @igor-sirotin .)

>   - `ChannelId`: the `channelId` passed to `createReliableChannel`.

Reviewer 2: [Followup Note] How is this created?

>   - Retrieval hint: the transport `MessageHash` of previous segments, exposed by the underlying [MESSAGING-API](/standards/application/messaging-api.md) upon ``MessageSendPropagatedEvent` reception.

Reviewer 2: ```suggestion
  - `Retrieval hint`: the transport `MessageHash` of previous segments, exposed by the underlying [MESSAGING-API](/standards/application/messaging-api.md) upon ``MessageSendPropagatedEvent` reception.
```

Author: Addressed in 76d8f2fe74772403fc1f9f9b624f02ee07728b2e

>   - Retrieval hint: the transport `MessageHash` of previous segments, exposed by the underlying [MESSAGING-API](/standards/application/messaging-api.md) upon ``MessageSendPropagatedEvent` reception.

Reviewer 2: [Dust]  `MessageHash` does not appear in the referenced spec - Which makes it hard to connect the concepts.

I'd use the case used in the spec to ensure there is no ambiguity 'message_hash`.
This also occurs with all the other identifiers which makes it difficult to find what part of the spec you are referring to

Author: Addressed in 59ff35b5e79e35fc5fc40fad51156df30a118591

Reviewer 2: Clean solution. In the future I think you could do this for all other identifiers too.

  The hint provider registered during [Node initialization](#node-initialization) performs this `MessageId → MessageHash` lookup. In turn, that mapping MUST be persisted by SDS using the `persistence` backend configured in `SdsConfig`.

> - Each sent segment is registered in an outgoing buffer.

Reviewer 2: [?] By which identifier? The random MessageID?

Author: In the current SDS implementation, the message is added completely to a queue. 
```nim
      channel.outgoingBuffer.add(
        UnacknowledgedMessage.init(message = msg, sendTime = getTime(), resendAttempts = 0)
      )
```

- The recipient sends acknowledgements back to the sender upon receiving segments.
- The sender removes acknowledged segments from the outgoing buffer.
- Unacknowledged segments are retransmitted after `acknowledgementTimeoutMs`.
- SDS state MUST be persisted using the `persistence` backend configured in `SdsConfig`.

### Rate Limit Manager

The Rate Limit Manager ensures compliance with [RLN](https://lip.logos.co/messaging/standards/core/17/rln-relay.html) rate constraints.

> - It tracks how many messages have been sent in the current epoch (only the first segment of each message counts toward the rate limit; subsequent segments are exempt).

Reviewer 2: > only the first segment of each message counts toward the rate limit; subsequent segments are exempt

[?] How does this work - how are some segments exempt? Are the segments sharing a RLN proof?

Author: This is incorrect. Great catch, thanks! Addressed in f4e5ed59a0e5ee0dc3b289e228c1c93c4bf371f1

Reviewer: Sounds similar to https://github.com/logos-messaging/specs/pull/104#discussion_r3131649106, but still slipped in here, good we fixed it 👍

> - When the limit is approached, segment dispatch MUST be delayed to the next epoch.

Reviewer 2: [Sand] SDS generates automatic delivery_receipt messages messages. What stops these from consuming the entire rate limit forever?

Author: The configured Rate Limits should support the required SDS messages.

Reviewer 2: Said another way: This seems like there is a potential for "Queue starvation". 

- If all messages are treated as equal priority
- If SDS delivery receipts are handled first / automatically

It appears that many messages could fill up the outbound queue, and prohibit an application from sending real messages.

> - The epoch size MUST match the `epochSizeMs` configured in `RateLimitConfig`.

Reviewer 2: [Pebble] I'm not sure what "epoch size" this is referring to? I don't see a parameter in 17/RLN-RELAY that corresponds to that.

Author: Good catch! Addressed in 94fab074d641608625f87136f5a27d4cc89bb064

Reviewer: @Ivansete-status I think `RateLimitConfig` shouldn't be here it all. We discussed it in https://github.com/logos-messaging/specs/pull/104#discussion_r3148303864

### Encryption Hook

The Encryption Hook provides a pluggable interface for upper layers to inject encryption.

- The hook is optional; when not provided, messages are sent unencrypted.

> - Encryption is applied per segment, after segmentation and SDS registration.

Reviewer 2: [Dust] What is `SDS Registration` I was confused for a long time as  Registration does not appear in 109/SDS. I see that you are referring to:

> Each sent segment is registered in an outgoing buffer.

Consider keeping the language plain so its easier to follow

Reviewer 2: >  Decryption is applied per segment, before SDS delivery.

[Dust] Reading this line, I think  `...  before being processed by SDS` better reflects your intention.

Author: Thanks!, addressed in 776a7db28ad77e84a8361b13593506830aa61c24

- Decryption is applied per segment, before SDS delivery.

> - The `Encryption` interface MUST be implemented by the caller.

Reviewer 2: [?] Am I correct in assuming that Callers must implement the  `Encryption` interface even if the encryption hook is not used because it is optional?

Author: Addressed in fd7e52dc5b4188b27044b37a62d61b84c20a0e07

   The hint provider converts an SDS `MessageId` into its corresponding `MessageHash`.
3. **Configure Segmentation persistence**: Supply the `Persistence` backend from `SegmentationConfig` to the
   [Segmentation](./segmentation.md) module so that partially reassembled messages survive restarts.
4. **Fetch missed messages**: Retrieve messages missed while offline as described in
   [MESSAGING-API — Fetching missed messages on startup](/standards/application/messaging-api.md#init-node-extended-definitions).

### Outgoing message processing

When `send` is called, the implementation MUST process `message` in the following order:

1. **Segment**: Split the payload into segments as defined in [SEGMENTATION](./segmentation.md). The maximum segment size MUST be reduced by the size of the SDS header added in step 2, so that each segment together with its SDS header stays within the network message size limit.
2. **Apply [SDS](https://lip.logos.co/ift-ts/raw/sds.html)** for each segment that is sent:
  - Generate a random MessageId. 40-char long. e.g. `a23f89b4df7e1ba5a709352da3a247a4fd609f4c`.
  - Persist the MessageId plus the payload for the current segment being sent.

>   - Get the retrieval hint from previously persisted messages (see point 5.).

Reviewer: These bullet points don't look "procedural", this is more of a parameters definition for SDS.

I'd rather move it under `## Components / ### SDS`, something like this:
```
SDS parameters used by this API are bound as follows:
- `MessageId`: a 40-character random hex string (e.g. `a23f89b4df7e1ba5a709352da3a247a4fd609f4c`), generated by the Reliable Channel API per segment.
- `ChannelId`: the `channelId` passed to `createReliableChannel`.
- Retrieval hint: the transport `MessageHash` of the segment, exposed by the underlying [MESSAGING-API](/standards/application/messaging-api.md). The hint provider registered during [Node initialization](#node-initialization) performs this `MessageId → MessageHash` lookup.
```

While _procedure_ describes the order of components, not how exactly we use them.

```suggestion
2. **Apply [SDS](https://lip.logos.co/ift-ts/raw/sds.html)**: register each segment in the SDS outgoing buffer (see [SDS](#scalable-data-sync-sds) for parameter bindings).
```

Author: Applied the suggestion in 81ee865e55ea525354c1ccb22ce04b3c3de690da

> 1. **Segment**: Split the payload into segments as defined in [SEGMENTATION](./segmentation.md). The maximum segment size MUST be reduced by the size of the SDS header added in step 2, so that each segment together with its SDS header stays within the network message size limit.

Reviewer 2: > reduced by the size of the SDS header added in step 2

[Sand] SDS does not use a header according to the spec. It uses a payload message with embedded bytes.

Author: In this case, `header` refers to anything that is not `content`, 
<img width="1524" height="456" alt="image" src="https://github.com/user-attachments/assets/d06dcba4-41b2-4fa2-af54-ab1537f5a8b2" />

Addressed in 9804f4ea7744596a73223afdf3bd1628343f87b9

Reviewer: While I believe it's important to keep transport messages within size constraints, I wonder how can we technically assure this.

@Ivansete-status what I think @jazzz mean here is that a "header" would be of fixed size. And it would be easy to consider it in segmentation.

But SDS has plenty of arbitrary length fields: strings, bytes, arrays:
```proto
message HistoryEntry {
  string message_id = 1; // Unique identifier of the SDS message, as defined in `Message`
  optional bytes retrieval_hint = 2; // Optional information to help remote parties retrieve this SDS message; For example, A Waku deterministic message hash or routing payload hash

  optional string sender_id = 3; // Participant ID of original message sender. Only populated if using optional SDS Repair extension
}

message Message {
  string sender_id = 1;           // Participant ID of the message sender
  string message_id = 2;          // Unique identifier of the message
  string channel_id = 3;          // Identifier of the channel to which the message belongs
  optional uint64 lamport_timestamp = 10;    // Logical timestamp for causal ordering in channel
  repeated HistoryEntry causal_history = 11;  // List of preceding message IDs that this message causally depends on. Generally 2 or 3 message IDs are included.
  optional bytes bloom_filter = 12;         // Bloom filter representing received message IDs in channel

  repeated HistoryEntry repair_request = 13; // Capped list of history entries missing from sender's causal history. Only populated if using the optional SDS Repair extension.

  optional bytes content = 20;             // Actual content of the message
}
```

And similar good point in https://github.com/logos-messaging/specs/pull/104#discussion_r3171484549 regarding encryption.

Author: Addressed in 74e81e0125e7b9122c2b4ab9464d2fbbe2fe4430

> 1. **Segment**: Split the payload into segments as defined in [SEGMENTATION](./segmentation.md). The maximum segment size MUST be reduced by the size of the SDS header added in step 2, so that each segment together with its SDS header stays within the network message size limit.

Reviewer 2: > so that each segment together with its SDS header stays within the network message size limit.

[?] What about the overhead from the ciphertext/encryption hook? Does that need to be considered?

Author: Yes indeed. This is an interesting point.
The current spec doesn't properly consider the encryption overhead within segments. And that raises the question whether we should perform such encryption as a very first step. This is something to consider in upcoming PRs in lips repo.

2. **Apply [SDS](https://lip.logos.co/ift-ts/raw/sds.html)**: register each segment in the SDS outgoing buffer (see [SDS](#scalable-data-sync-sds) for parameter bindings).

> 3. **Encrypt**: If an `Encryption` implementation is provided, encrypt each segment before transmission.

Reviewer 2: https://github.com/logos-messaging/specs/pull/104/changes#diff-12c6309e444a295df51736d66bcfd259907dbda2b65ffcceb13ec8d71b2abcf1R133

> - The `Encryption` interface MUST be implemented by the caller.

Does that mean that segments are always encrypted?

Author: Yes, according to the current spec. We might need to revisit that once we move to lips.

> 3. **Encrypt**: If an `Encryption` implementation is provided, encrypt each segment before transmission.

Reviewer 2: [Dust] Referring to multiple payload types as a 'segment' without defining expressly what you mean by segment is hard to follow. 

I would assume that a segment is the output from segmentation, however its being used to refer to the output of SDS. I'd either define what you mean by segment so readers are aware; or use a different term such as SDS:Message to be clear which type you are referring to

Author: Addressed in f3ba3f6d3fa6ac73d53093efcb4c1cbebaaedc53

  - Update the persisted MessageId entry, in point 2., with the sent MessageHash, only when

>     the event `MessageSendPropagatedEvent` is triggered (that event is defined in [MESSAGING-API](/standards/application/messaging-api.md)).

Reviewer: Same as https://github.com/logos-messaging/specs/pull/104/changes#r3161712542 here.
This is more of "how we use SDS" that a procedural thing.

Maybe add something like this:
```
The hint provider maintains a `MessageId → MessageHash` mapping,
populated on `MessageSendPropagatedEvent` and used to resolve retrieval hints for outgoing segments.
This mapping MUST be persisted using the `persistence` backend configured in `SdsConfig`.
```

Then this procedure can be kept as simple as:
```suggestion
5. **Dispatch**: Send each segment via the underlying [MESSAGING-API](/standards/application/messaging-api.md).
```

Author: Addressed in 0b3a4aeb5ce1bb3894855f8494b8ceaa859efcd5 and 8483bb10d9ee60e05c12fc68ef5670e7a24f2640

### Incoming message processing

When a segment is received from the network, the implementation MUST process it in the following order:

1. **Decrypt**: If an `Encryption` implementation is provided, decrypt the segment.
2. **Apply [SDS](https://lip.logos.co/ift-ts/raw/sds.html)**: Deliver the segment to the SDS layer, which emits acknowledgements and detects gaps.

>    - **Detect missing dependencies and fetch from store**: If SDS detects a gap in the causal history (i.e., a referenced predecessor segment has not yet been received), and considers the message is irretrievably lost, 

Reviewer 2: >  If SDS detects a gap in the causal history (i.e., a referenced predecessor segment has not yet been received)

The clarifying statement is great. I immediately understand what is meant.

> If SDS detects a gap in the causal history

Nitpicky: The casual history (list of SDS:HistoryEntries) would not have a Gap. The data structure is intact, the client is just missing a message.

[Dust] I would lead with the clear part. 

" If SDS detects that a referenced predecessor segment has not yet been received,"
or simplify to:
" if SDS detects a message in the causal history which has not yet been received "
or 
" if SDS detects that a message has been lost, and it cannot recover it"

Author: Addressed in this, thanks! c7eb2bef623d183d0419f32c947a5b0cdb3ea555

>    the implementation MUST attempt to retrieve the missing segment throughout store protocol (store API is never exposed and hence, store queries are handled internally).

Reviewer 2: [Dust] Remove the word 'irretrievably' as it conflicts with the next line, where the implementation MUST retrieve 

**Irretrievable**: describes something that cannot be retrieved, recovered, regained, repaired, or undone. It signifies a permanent loss, finality, or irreparable damage. - [Definition](https://www.merriam-webster.com/dictionary/irretrievable)

If it cannot be retrieved, I'd ask why then why does the implementation need to attempt to retrieve it?

Focusing on the distinction that SDS cannot recover it would be helpful, and highlight that this is an advanced recovery mechanism.

Author: Addressed in 895a00c5ebe77efce01c0ede49cb5c4646111a07

>    the implementation MUST attempt to retrieve the missing segment throughout store protocol (store API is never exposed and hence, store queries are handled internally).

Reviewer 2: [?] So the strategy for Store nodes is that the API makes it difficult to use to avoid abuse, however its still available if developers want to implement their own spec compliant API and access it?

Author: No, the following explains it better: 5909a2f48df03dbe2bd821f165835c149bf390d9

>    the implementation MUST attempt to retrieve the missing segment throughout store protocol (store API is never exposed and hence, store queries are handled internally).

Reviewer 2: [Sand] It would be helpful to explain, that this is what the MessageHash is for. 

The ``MessageId → MessageHash`` mapping is never really explained

Author: Addressed in ac4883fa15c21816a0a962a3eeee9ea6a3f5dac8

3. **Reassemble**: Once all segments for a message have been received, reassemble and emit a `reliable:message:received` event.

### Rate limiting

> When `RateLimitConfig.enabled` is `true`, the implementation MUST space segment transmissions

Reviewer 2: [?] Is there a use case where you don't want RateLimiting? If all networks eventually use RLN, disabling Local Rate limiting would cause developers/users to  fully disclose SK

Author: Yes, certain implementations may opt to not use RLN if they consider it is not mature enough.

      - name: contentTopic
        type: string
        description: "The topic this channel listens and sends on. This has routing and filtering connotations."
      - name: senderId
        type: string
        description: "An identifier for this sender. SHOULD be unique and persisted between sessions."
      - name: encryption
        type: optional<Encryption>
        default: none
        description: "Optional pluggable encryption implementation. If none, messages are sent unencrypted."
    returns:
      type: result<ReliableChannel, error>

  closeChannel:
    description: "Closes a reliable channel, releases all associated resources and internal state,
    and unsubscribes from its content topic via the underlying [MESSAGING-API](/standards/application/messaging-api.md)."
    parameters:
      - name: channel
        type: ReliableChannel
        description: "The channel handle returned by `createReliableChannel`."
    returns:
      type: result<void, error>
```

### Channel usage

```yaml
functions:
  send:
    description: "Send a message through a reliable channel. The message is always segmented,

>     SDS-tracked, rate-limited, and encrypted (if configured)."

Reviewer 2: > The message is always segmented,
    SDS-tracked, rate-limited, and encrypted (if configured)."

There is ambiguity if the `(If Configured)` only applies. to `encrypted` or if it applies to all the items. To me it reads "only encryption is optional"

However:

> When `RateLimitConfig.enabled` is `true`,
Implies that Ratelimiting can be disabled

either:
[Dust] Apply ("If configured") to all the elements that can be disabled
or:
[Dust] Suggestion: "The message is always processed in the same order (segmented,
    SDS-tracked, rate-limited, and encrypted)."

depending on your intention

Author: Addressed in https://github.com/logos-messaging/specs/commit/a26a7992a61b247f70a18566ed0b152f97cc5836

    parameters:
      - name: channel
        type: ReliableChannel
        description: "The channel handle returned by `createReliableChannel`."
      - name: message
        type: array<byte>
        description: "The raw message payload to send."
    returns:
      type: result<RequestId, error>

>       description: "Returns a `RequestId` that callers can use to correlate subsequent `MessageSentEvent` or `MessageSendErrorEvent` events."

Reviewer 2: Returning a random request_id would be annoying, in many cases. Developers almost guaranteed have a `message_id` for the object they are passing in.

[Dust] Allow developers to pass in their own identifier so they don't need to persist a mapping between RequestID and their messageID

Author: Yes good point but I'm afraid this is needed, i.e., we need to relate one `MessageId` with possible many `RequestId`, i.e., it may need to send the same message twice or more times, in which case, each will generate a different `RequestId`.

Reviewer 2: This is a very helpful distinction!

```

### Node configuration

This spec extends `NodeConfig`, needed to create a node, which is 
defined in [MESSAGING-API](/standards/application/messaging-api.md), 
with `sds_config` and `rate_limit_config` fields.

```yaml
NodeConfig:  # Extends NodeConfig defined in MESSAGING-API
  fields:
    sds_config:
      type: SdsConfig
      description: "SDS configuration. See SdsConfig defined in this spec."
```

### Type definitions

```yaml
types:

  ReliableChannel:
    type: object
    description: "A handle representing an open reliable channel.
    Returned by `createReliableChannel` and used to send messages and receive events.

>     Internal state (SDS, segmentation, encryption) is managed by the implementation."

Reviewer 2: [?] What internal state is there for encryption?

> The separation between Reliable Channels and encryption ensures the API remains **agnostic to identity and key management concerns,**
which are handled by higher layers.

Author: That's open for the implementation. The state for encryption may be nothing; encryption keys; encryption state maybe needed by De-MLS; etc.

Reviewer 2: > is managed by the implementation.

I"m still confused:  Who is the implementation here? I assumed it was SDK/API providing the reliable channel API (ie: Not the application)

- I thought applications were responsible for encryption
- I thought the "implementation" was responsible for SDS and segmentation.

The statement implies that everything is managed by the same entity

    events:

>       "reliable:message:received":

Reviewer 2: [Sand] As a developer I am interested in being notified of a messages that were being received however a error occurred.  ie: MessageReceiveErrorEvent

Author: Sorry I don't fully get the comment

Reviewer 2: Currently there are 5 events:

- MessageSentEvent
- MessageSendErrorEvent

- MessageDeliveredEvent
- MessageDeliveryErrorEvent

- MessageReceivedEvent

My assumption is that during the receiving process its possible for an exception to be emitted. e.g. Malformed packet, bad decryption, state conflict, etc.

Adding 
- MessageReceiveErrorEvent to notify developers that an error has occurred while processing inbound messages would help them discover errors and lost messages

Author: Ah yes, good point. We are describing the errors from the sender PoV mostly. I think we might need that type of error as we evolve the implementation. Then I suggest to adapt the spec once that occurs.

    For end-to-end confirmation, listen for `MessageDeliveredEvent`."
    fields:
      requestId:
        type: RequestId
        description: "The identifier of the `send` operation whose segments have all been dispatched to the network."

  MessageDeliveredEvent:
    type: object
    description: "Event emitted when the recipient has confirmed end-to-end receipt of a message via SDS acknowledgements.
    This event is fired asynchronously after `MessageSentEvent`, once the SDS layer receives explicit acknowledgements from the recipient."
    fields:
      requestId:
        type: RequestId
        description: "The identifier of the `send` operation confirmed as delivered by the recipient."

  MessageSendErrorEvent:
    type: object
    description: "Event emitted when one or more segments of a message could not be dispatched to the network.
    This indicates a network-level failure; the message was never fully transmitted."
    fields:
      requestId:
        type: RequestId
        description: "The identifier of the `send` operation that failed to dispatch."
      error:
        type: string

>         description: "Human-readable description of the dispatch failure."

Reviewer 2: [?] In many applications preparing a "Message" results in a permanent state change. 

Is the expectation that Developers need to retain a copy of this message, until it has been delivered successfully?

Author: The apps should always keep their own copies of messages and keep an state for each of them. I think that's something will happen naturally.

I would use the following message states in an app:
- Sent
- PendingConfirmation
- ConfirmedReceived
- PersistedLocally
- PersistedRemotely
- ...

But any app may implement their own states.

  MessageDeliveryErrorEvent:
    type: object
    description: "Event emitted when end-to-end delivery could not be confirmed.

>     Fired after `maxRetransmissions` attempts have been exhausted without receiving an SDS acknowledgement from the recipient."

Reviewer 2: [?] As a developer what does this Error actually tell me?

- Does this mean that the message definitely won't arrive?
- Does this mean that it might arrive, but we don't know? Which seems like the default case.

does this mean I should resend? or do I risk duplicate messages if I do?

Author: This tells that the message was not confirmed by the receiver but the network processed it properly. Clarifying in 4d539e0eb34ee096b9819f9fa7068ed2606293f4

Reviewer 2: > The message reached the network and there's no need to explicit re-send.

This is much clearer! My message will still get there, its just taking some time.

  RateLimitConfig:
    type: object
    description: Rate limiting configuration, containing RLN-specific attributes.
    fields:
      enabled:
        type: bool
        default: false
        description: "Whether rate limiting is enforced. SHOULD be true when RLN is active."
      epochSizeMs:
        type: uint
        default: 600000  # 10 minutes
        description: "The epoch size used by the RLN relay, in milliseconds."

  Encryption:
    type: object
    description: "Interface for a pluggable encryption mechanism.
    When provided as a parameter to `createReliableChannel`, the API consumer MUST implement both encrypt and decrypt operations.
    Implementations MAY use different signatures than those described below, as long as each operation accepts a byte array and returns a byte array."
    fields:
      encrypt:
        type: function
        description: "Encrypts a byte payload. Returns the encrypted payload."
        parameters:
          - name: content
            type: array<byte>
        returns:

>           type: result<array<byte>, error>

Reviewer 2: [Sand] What error type is returned from the encryption function?

Author: Same. Let's enhance once in lips.

      decrypt:
        type: function
        description: "Decrypts a byte payload. Returns the decrypted payload."
        parameters:
          - name: payload
            type: array<byte>
        returns:

>           type: result<array<byte>, error>

Reviewer 2: [Pebble] These functions don't provide any mechanism for developers to know how a message was encrypted, or how to decrypt it. 

Assuming an best case (that developers use a single static key per channel) A developer would have to iteratively attempt decryption with each key they have. In the chat usecase we are assuming that the average user has 200 chats, resulting in 200 attempts per message. 

If developers choose an encryption scheme that is much safer, they could reasonably have thousands of keys at a given time.

Author: That's a very important point. We need to tackled that in a separate PR because that would require more work. The current `Encryption` mechanism is too weak and stateless, which is too naïve.

>           type: result<array<byte>, error>

Reviewer 2: [Pebble] How are developers supposed to determine if decryption was successful?

If all that is receive is a byte array, and there are  multiple encryption keys (because I have multiple reliable channels) - How do I determine I found the right key? 

I as the developer have no idea what the payload structure should be because its generated internally. Each decryption will result in a seemingly random series of bytes, If I don't know what I should be expecting.

Author: We will tackle that in a separate PR, in lips

    * [Scalable Data Sync (SDS)](#scalable-data-sync-sds)
    * [Rate Limit Manager](#rate-limit-manager)
    * [Encryption Hook](#encryption-hook)
  * [Security/Privacy Considerations](#securityprivacy-considerations)
  * [Copyright](#copyright)
<!-- TOC -->

## Abstract

This document specifies the **Reliable Channel API**,
an application-level interface that sits between the application layer and the [MESSAGING-API](/standards/application/messaging-api.md) plus [P2P-RELIABILITY](/standards/application/p2p-reliability.md), i.e., `application` <-> **reliable-channel-api** <-> `messaging-api/p2p-reliability`.

It bundles segmentation, end-to-end reliability via [Scalable Data Sync (SDS)](https://lip.logos.co/ift-ts/raw/sds.html), rate limit management, and a pluggable encryption hook
into a single interface for sending and receiving messages reliably.

## Motivation

The [MESSAGING-API](/standards/application/messaging-api.md) provides peer-to-peer reliability via [P2P-RELIABILITY](/standards/application/p2p-reliability.md),
but does not provide high end-to-end delivery guarantees from sender to recipient.

This API addresses that gap by introducing:

> - **[SEGMENTATION-API](/standards/application/segmentation.md)** to handle large messages exceeding network size limits.

Reviewer: I think I commented about renaming the path before. But I missed that the link text should also be renamed of course. There's not such thing as "segmentation api".
```suggestion
- **[SEGMENTATION](/standards/application/segmentation.md)** to handle large messages exceeding network size limits.
```

Author: Addressed in 682e3525e22598c1a59a61916274da1c4ffbddea

2. **Apply [SDS](https://lip.logos.co/ift-ts/raw/sds.html)**: Register each segment with the SDS layer to track acknowledgements and enable retransmission.
3. **Encrypt**: If an `Encryption` implementation is provided, encrypt each segment before transmission.
4. **Rate Limit**: If `RateLimitConfig.enabled` is `true`, delay dispatch as needed to comply with [RLN](https://lip.logos.co/messaging/standards/core/17/rln-relay.html) epoch constraints.
5. **Dispatch**: Send each segment via the underlying [MESSAGING-API](/standards/application/messaging-api.md).

### Incoming message processing

When a segment is received from the network, the implementation MUST process it in the following order:

1. **Decrypt**: If an `Encryption` implementation is provided, decrypt the segment.
2. **Apply [SDS](https://lip.logos.co/ift-ts/raw/sds.html)**: Deliver the segment to the SDS layer, which emits acknowledgements and detects gaps.
   - **2.1. Detect missing dependencies**: If SDS detects a gap in the causal history (i.e., a referenced predecessor segment has not yet been received), the implementation MUST attempt to retrieve the missing segment.

>    - **2.2. Fetch from store**: The implementation MUST provide a mechanism to extract a retrieval hint from the received segment (e.g., a store cursor or message hash embedded in the SDS causal history). The missing segment MUST then be fetched from a store node via the [MESSAGING-API](/standards/application/messaging-api.md) store query, and re-injected into the incoming processing pipeline from step 1.

Reviewer: In `Outgoing message processing` it should also set the retrieval hints

Author: Sorry, I can't quite follow the question.
Isn't that already covered with the following phrase?

> 2. **Apply [SDS](https://lip.logos.co/ift-ts/raw/sds.html)**: Register each segment with the SDS layer to track acknowledgements and enable retransmission.

Reviewer: @Ivansete-status Pardon, let me clarify it. 

SDS provides a way to pass an arbitrary "retrieval hint". Reliable Channel API knows that the underlying transport service is Logos Delivery. So Reliable Channels should pass `RequestID` (from Messaging API) as the retrieval hints for SDS.

It might be obvious for us, but it should be part of this spec, because it's the one that brings SDS and Logos Delivery together.

Author: I think this gathers your recommendations in our today's sesstion: 6ff41e97f978eefabbf6e8d9c99700a80ef0df93

We are persisting the SDS hint (`MessageHash`) after we get the confirmation that a certain segment reached a store node.

>    - **2.2. Fetch from store**: The implementation MUST provide a mechanism to extract a retrieval hint from the received segment (e.g., a store cursor or message hash embedded in the SDS causal history). The missing segment MUST then be fetched from a store node via the [MESSAGING-API](/standards/application/messaging-api.md) store query, and re-injected into the incoming processing pipeline from step 1.

Reviewer: Looks like 2.2. and 2.1. are about the same thing, no?
2.1. should be done using retrieval hints. And retrieval hints shouldn't be exposed to the user.

Author: Oh yes, too verbose. Addressed in 76ccd4154914869d476fa7252184cce57a8f3190

        type: ReliableSendId
        description: "The identifier of the `send` operation confirmed as delivered by the recipient."

  MessageSendErrorEvent:
    type: object
    description: "Event emitted when a message send operation fails after exhausting retransmission attempts."
    fields:
      requestId:
        type: ReliableSendId
        description: "The identifier of the `send` operation that failed after exhausting retransmission attempts."
      error:
        type: string
        description: "Error message describing what went wrong"

  ReliableSendId:
    type: string
    description: "Unique identifier for a single `send` operation on a reliable channel.
    It groups all segments produced by segmenting one message, so callers can correlate
    acknowledgement and error events back to the original send call.
    Internally, each segment is dispatched as an independent [MESSAGING-API](/standards/application/messaging-api.md) call,
    producing one `RequestId` (as defined in [MESSAGING-API](/standards/application/messaging-api.md)) per segment.
    A single `ReliableSendId` therefore maps to one or more underlying `RequestId` values,
    one per segment sent."

> 

Reviewer: I think we've had `SegmentationConfig` here before 🤔 
The reason I ask is to suggest to specify default `parityRate: 0`, this will disable Reed-Solomon, as discussed in https://github.com/logos-messaging/specs/pull/91#discussion_r3152634622

Author: Addressed in a23f89b4df7e1ba5a709352da3a247a4fd609f4c

I think is better to make the flag more explicit, i.e., `enableReedSolomon`.
We might need to adapt the segmentation spec to allow that optional behaviour.

    * [Messaging](#messaging)
      * [Messaging type definitions](#messaging-type-definitions)
      * [Messaging function definitions](#messaging-function-definitions)
      * [Messaging extended definitions](#messaging-extended-definitions)
  * [Components](#components)
    * [Segmentation](#segmentation)
    * [Scalable Data Sync (SDS)](#scalable-data-sync-sds)
    * [Rate Limit Manager](#rate-limit-manager)
    * [Encryption Hook](#encryption-hook)
  * [Security/Privacy Considerations](#securityprivacy-considerations)
  * [Copyright](#copyright)
<!-- TOC -->

## Abstract

This document specifies the **Reliable Channel API**,
an application-level interface that sits between the Logos Chat and the [MESSAGING-API](/standards/application/messaging-api.md) plus [P2P-RELIABILITY](/standards/application/p2p-reliability.md), i.e., `logos-chat` <-> **reliable-channel-api** <-> `messaging-api/p2p-reliability`.

It bundles segmentation, end-to-end reliability via [Scalable Data Sync (SDS)](https://lip.logos.co/ift-ts/raw/sds.html), rate limit management, and a pluggable encryption hook
into a single interface for sending and receiving messages reliably.

> Application developers SHOULD use this API when end-to-end reliability across multiple routing hops is required.

Reviewer 2: [Dust] Use of 'Should' here implies this is a protocol requirement - I recommend using plain language in the abstract.

1. Abstract SHOULD contain a summary of the specification, rather than defining protocol requirements - those should come in sections below.
2. What application developers do seems out of scope for this document. I would focus on why/when this protocol is a good choice for developers, rather than softly requiring that they use it.

Author: Good point. I remove this comment completely to avoid confusion, in https://github.com/logos-messaging/specs/pull/104/changes/b44c2a5c27d4ae7261ad171cfc22934653b08f61

## Motivation

The [MESSAGING-API](/standards/application/messaging-api.md) provides peer-to-peer reliability via [P2P-RELIABILITY](/standards/application/p2p-reliability.md),
but does not provide high end-to-end delivery guarantees from sender to recipient.

This API addresses that gap by introducing:
- **Segmentation** to handle large messages exceeding network size limits.
- **SDS** to provide causal-history-based end-to-end acknowledgement and retransmission.
- **Rate Limit Manager** to comply with [RLN](https://lip.logos.co/messaging/standards/core/17/rln-relay.html) constraints when sending segmented messages.
- **Encryption Hook** to allow upper layers (e.g., the Logos Chat) to provide a pluggable encryption mechanism.

> 

Reviewer 2: [Pebble] I don't follow the motivation for  behind `encryption`. Its easy to make assumptions about the goals here, which makes it difficult to understand if the approach is effective.

Providing clarity on why encryption is providing would be helpful.

Author: Good point! The following aims to address that: ffea66cacc92538c0617b80602a3efa334df996a

The separation between Reliable Channels and encryption ensures the API remains agnostic to identity and key management concerns,
which are handled by higher layers.

## Syntax

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT",
"RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC2119](https://www.ietf.org/rfc/rfc2119.txt).

> ## API design

Reviewer 2: [Pebble] Its hard to understand exactly what is happening, when the focus is on the API  

This Spec highlights the API of reliable channels and does a great job of describing those functions. However many of the details of the mechanics of reliable channels are not discussed or hard to find. 

Those are the details I am specifically focused on. 

Id suggest either:
1. having two documents one that focused on reliableChannels as a protocol, and one that highlights how your team wants to implement it. 
2. Define the ReliableChannel specification, and add the API as an implementation suggestion.

Author: I think I like option 2. Feel free to push anything that you consider interesting. Or if you prefer having a chat before pushing any change, you can submit a PR to this `add-new-reliability-spec` branch and we can comment there. I'm fine in either case.

Reviewer: As in https://github.com/logos-messaging/specs/pull/89, I think it would be better to put define the architecture and goals first, and only then propose the API. This will change the focus from code to goals, and simplify reading.

I.e. first: architecture, procedures, routines (you have it all in the end), then functions/types.

Author: Addressed in 137f22f16d7652c08cdebb4c54fa33fd430f77fb

            type: array<byte>
        returns:
          type: result<array<byte>, error>
      decrypt:
        type: function
        description: "Decrypts a byte payload. Returns the decrypted payload."
        parameters:
          - name: ciphertext
            type: array<byte>
        returns:
          type: result<array<byte>, error>
```

### Channel

#### Channel type definitions

```yaml
types:
  ReliableEnvelope:
    type: object
    description: "Represents an extension of the MessageEnvelope defined in MESSAGING-API aiming mostly to allow
    including a reliable channel id attribute that, when given, it will assume the message should be sent reliably. Therefore, we leave open for implementors to replace the MessageEnvelope implementation and just add the reliable channel id there."
    fields:
      messageEnvelope:
        type: MessageEnvelope
        description: "Please refer to the [MESSAGING-API](/standards/application/messaging-api.md) for details."
      channelId:
        type: string
        default: ""
        description: "Reliable channel identifier.
        If empty, the message is considered ephemeral and is not sent reliably.

>         If != empty, the message will be segmented, SDS'ed and encrypted under the given reliable channel."

Reviewer 2: Its not clear to me How encryption fits in here. Is the whole payload encrypted?

Author: Everything under messageEnvelop is encrypted. The clarification is added in: https://github.com/logos-messaging/specs/pull/104/changes/42316c9e7e14865113a936a609479192f55429c1

3. **Reassemble**: Once all chunks for a message have been received, reassemble and emit a `reliable:message:received` event.

**Retransmission**:

The SDS layer MUST retransmit unacknowledged chunks after `acknowledgementTimeoutMs`,
up to `maxRetransmissions` times.
If a chunk is not acknowledged after all retransmission attempts, a `reliable:message:send-error` event MUST be emitted.

**Rate limiting**:

When `RateLimitConfig.enabled` is `true`, the implementation MUST space chunk transmissions
to comply with the RLN epoch constraints defined in `epochSizeMs`.
Chunks MUST NOT be sent at a rate that would violate the RLN message rate limit for the active epoch.

## Components

### Segmentation

Segmentation splits outgoing messages into chunks before SDS processing.
This ensures each chunk fits within the network's maximum message size.

- The default chunk size is `100 KiB` (`102400 bytes`.)
- Chunks MUST be tagged with metadata sufficient for the receiver to reassemble the original message:
  a message identifier, chunk index, and total chunk count.
- Segmentation MUST be transparent to the SDS layer; SDS operates on individual chunks.

> - The maximum allowed message size is `1MiB` (`1048576 bytes`.)

Reviewer 2: Where does this Maximum value come from?

Author: This is based on the current max size used by _Status_.

  * [Syntax](#syntax)
  * [API design](#api-design)
    * [IDL](#idl)
    * [Architectural position](#architectural-position)
  * [The Reliable Channel API](#the-reliable-channel-api)
    * [Common](#common)
      * [Common type definitions](#common-type-definitions)
    * [Channel](#channel)
      * [Channel type definitions](#channel-type-definitions)
      * [Channel function definitions](#channel-function-definitions)
      * [Channel predefined values](#channel-predefined-values)
      * [Channel extended definitions](#channel-extended-definitions)
    * [Messaging](#messaging)
      * [Messaging type definitions](#messaging-type-definitions)
      * [Messaging function definitions](#messaging-function-definitions)
      * [Messaging extended definitions](#messaging-extended-definitions)
  * [Components](#components)
    * [Segmentation](#segmentation)
    * [Scalable Data Sync (SDS)](#scalable-data-sync-sds)
    * [Rate Limit Manager](#rate-limit-manager)
    * [Encryption Hook](#encryption-hook)
  * [Security/Privacy Considerations](#securityprivacy-considerations)
  * [Copyright](#copyright)
<!-- TOC -->

## Abstract

This document specifies the **Reliable Channel API**,

> an application-level interface that sits between the Chat SDK and the [MESSAGING-API](/standards/application/messaging-api.md) plus [P2P-RELIABILITY](/standards/application/p2p-reliability.md), i.e., `chat-sdk` <-> **reliable-channel-api** <-> `messaging-api/p2p-reliability`.

Reviewer 2: For consistency we use `LogosChat` instead of `chatSDK`

Author: Addressed in: https://github.com/logos-messaging/specs/pull/104/changes/fa7ff625c225fcad3df8baf3ecb28ea59c8d0de8

    type: object
    description: "Represents the minimum unit of data transmitted across the network. Its size is governed by SegmentationConfig. Chunks are chained to their predecessors, and any chunk whose linkage to its predecessor cannot be verified MUST be discarded.
    Only the first chunk on every message, is rate-limited (RLN) and the subsequent chunks MUST be accepted rate-limit wise."
    fields:
      chunkId:
        type: string
        description: UUID for the chunk
      previousChunkId:
        type: string
        description: Previous chunk ID. Chunks with unvalidated previous chunks MUST be rejected.
        This field is empty when chunkIndex == 0.
      chunkIndex:
        type: uint
        description: "Zero-based position of this chunk within the ordered sequence of chunks that form the original message."
      lastChunkIndex:
        type: uint
        description: "All chunks of same message have the same value, i.e., num-chunks - 1."
      requestId:
        type: RequestId
        description: "The request id this chunk belongs to. This is generated by the send function and its type is defined in messaging-api."

  IEncryption:
    type: object

>     description: "Interface for a pluggable encryption mechanism. Implementors MUST provide encrypt and decrypt operations."

Reviewer 2: What the reasoning behind making encryption a requirement? Is the expectation that developers could provide a NOOP for public reliability?

Author: Addressed in https://github.com/logos-messaging/specs/pull/104/changes/92a746cca864f61a1978ac7bf707e82e33732b16

Reviewer 2: I see that IEncryption is not 'Optional' that makes sense - however this MUST is still confusing to me. 

What exactly are you trying to convey with this statement?

Author: I changed the description in this: 638f8676d65f8dc82dce67253f67e80819dfbd00
Let me know if still unclear.

    fields:
      encrypt:
        type: function
        description: "Encrypts a byte payload. Returns the encrypted payload."
        parameters:
          - name: plaintext
            type: array<byte>
          - name: encryptionKey

>             type: array<byte>

Reviewer 2: [Pebble] Passing the encryption key to the encryption callback appears to break abstraction levels.

- Encryption key types are often coupled with an encryption mechanism. Which key types will Reliable channels support?
- If a keytype can only support a subset of encryption schemes, what is the value in the callback?

Author: Addressed in https://github.com/logos-messaging/specs/pull/104/changes/9725dc84adc61e4f929c495a9cbecde2e174ca66

        parameters:
          - name: ciphertext
            type: array<byte>
          - name: encryptionKey
            type: array<byte>
        returns:
          type: result<array<byte>, error>
```

### Channel

#### Channel type definitions

```yaml
types:
  ReliableEnvelope:
    type: object
    description: "Represents an extension of the MessageEnvelope defined in MESSAGING-API aiming mostly to allow
    including a reliable channel id attribute that, when given, it will assume the message should be sent reliably. Therefore, we leave open for implementors to replace the MessageEnvelope implementation and just add the reliable channel id there."
    fields:
      messageEnvelope:
        type: MessageEnvelope
        description: "Please refer to the [MESSAGING-API](/standards/application/messaging-api.md) for details."
      channelId:
        type: string
        default: ""
        description: "Reliable channel identifier.
        If empty, the message is considered ephemeral and is not sent reliably.
        If != empty, the message will be segmented, SDS'ed and encrypted under the given reliable channel."

  ReliableChannel:
    type: object

>     description: "A reliable channel instance or communication group having end-to-end reliable messaging on a single content topic."

Reviewer 2: What is a "communication group" exactly? could use a definition somewhere

Author: Simplified in: https://github.com/logos-messaging/specs/pull/104/changes/27fef43f95a64a9d5204de412c02eec77823eada

```yaml
values:

  DefaultSegmentationConfig:
    type: SegmentationConfig
    fields:
      chunkSizeBytes: 102400  # 100 KiB

  DefaultSdsConfig:
    type: SdsConfig
    fields:
      historyBackend: "memory"
      acknowledgementTimeoutMs: 5000
      maxRetransmissions: 5
      causalHistorySize: 2

  DefaultRateLimitConfig:
    type: RateLimitConfig
    fields:
      enabled: true
      epochSizeMs: 600000
```

#### Channel extended definitions

**State management**:

Each `ReliableChannel` MUST maintain internal state for SDS, including:
- A committed message log recording sent messages and their acknowledgement status.
- An outgoing buffer of unacknowledged message chunks pending confirmation.
- An incoming buffer for partially received segmented messages awaiting full reassembly.

The state MUST be persisted according to the `historyBackend` specified in `SdsConfig`.
The default `memory` backend does not survive process restarts; implementors providing persistence SHOULD use a durable backend.

**Encryption**:

The `encryption` field in `ReliableChannelConfig` is intentionally optional.
The Reliable Channel API is agnostic to encryption mechanisms.

> Encryption is considered a concern of upper layers (e.g., the Chat SDK).

Reviewer 2: [Sand] This doesn't seem to rely on LogosChat - consider making it application agnostic

Author: Addressed in https://github.com/logos-messaging/specs/pull/104/changes/c3daa9d127fa4123cb1d1469ffcb35f20d074dd6

- The sender removes acknowledged chunks from the outgoing buffer.
- Unacknowledged chunks are retransmitted after `acknowledgementTimeoutMs`.
- SDS state MUST be persisted in the configured `historyBackend`.

### Rate Limit Manager

The Rate Limit Manager ensures compliance with [RLN](https://lip.logos.co/messaging/standards/core/17/rln-relay.html) rate constraints.

- It tracks how many messages have been sent in the current epoch (only the first chunk of each message counts toward the rate limit; subsequent chunks are exempt).
- When the limit is approached, chunk dispatch MUST be delayed to the next epoch.
- The epoch size MUST match the `epochSizeMs` configured in `RateLimitConfig`.

### Encryption Hook

The Encryption Hook provides a pluggable interface for upper layers to inject encryption.

- The hook is optional; when not provided, messages are sent as plaintext.
- Encryption is applied per chunk, after segmentation and SDS registration.
- Decryption is applied per chunk, before SDS delivery.
- The `IEncryption` interface MUST be implemented by the caller (e.g., the Chat SDK).
- The Reliable Channel API MUST NOT impose any specific encryption scheme.

> ## Security/Privacy Considerations

Reviewer 2: SDS Contains a senderID, depending on how this will be implemented there is a potential privacy issue with unencrypted public reliability. 

[Dust] Add privacy implications for unencrypted reliability

Author: Good point! Should we always encrypt the `channelId` to avoid privacy concerns? We can have a simple/fix encryption for that, without ratcheting in this particular case. wdyt?

Reviewer 2: > Should we always encrypt the channelId to avoid privacy concerns?

- Who would have access to this key? 
- How do you get the key to them?

Answering these questions has a major impact on UX. If it work for your use case, its easy enough to derive an encryption keym but depends heavily on the required security level

  * [Syntax](#syntax)
  * [API design](#api-design)
    * [IDL](#idl)
    * [Architectural position](#architectural-position)
  * [The Reliable Channel API](#the-reliable-channel-api)
    * [Common](#common)
      * [Common type definitions](#common-type-definitions)
    * [Channel](#channel)
      * [Channel type definitions](#channel-type-definitions)
      * [Channel function definitions](#channel-function-definitions)
      * [Channel predefined values](#channel-predefined-values)
      * [Channel extended definitions](#channel-extended-definitions)
    * [Messaging](#messaging)
      * [Messaging type definitions](#messaging-type-definitions)
      * [Messaging function definitions](#messaging-function-definitions)
      * [Messaging extended definitions](#messaging-extended-definitions)
  * [Components](#components)
    * [Segmentation](#segmentation)
    * [Scalable Data Sync (SDS)](#scalable-data-sync-sds)
    * [Rate Limit Manager](#rate-limit-manager)
    * [Encryption Hook](#encryption-hook)
  * [Security/Privacy Considerations](#securityprivacy-considerations)
  * [Copyright](#copyright)
<!-- TOC -->

## Abstract

This document specifies the **Reliable Channel API**,

> an application-level interface that sits between the Logos Chat and the [MESSAGING-API](/standards/application/messaging-api.md) plus [P2P-RELIABILITY](/standards/application/p2p-reliability.md), i.e., `logos-chat` <-> **reliable-channel-api** <-> `messaging-api/p2p-reliability`.

Reviewer: This specification should not mention `Logos Chat` at all. While Reliable Channel API can/will be used by Logos Chat, we don't solely implement the API for it. Otherwise we'd just implement thid functionality directly in `logos-chat` 🙂

Author: Addressed in b2ecac4ad5215ce6b9041a9214045e3a9245cf27

It bundles segmentation, end-to-end reliability via [Scalable Data Sync (SDS)](https://lip.logos.co/ift-ts/raw/sds.html), rate limit management, and a pluggable encryption hook
into a single interface for sending and receiving messages reliably.

## Motivation

The [MESSAGING-API](/standards/application/messaging-api.md) provides peer-to-peer reliability via [P2P-RELIABILITY](/standards/application/p2p-reliability.md),
but does not provide high end-to-end delivery guarantees from sender to recipient.

This API addresses that gap by introducing:
- **Segmentation** to handle large messages exceeding network size limits.
- **SDS** to provide causal-history-based end-to-end acknowledgement and retransmission.

> - **Rate Limit Manager** to comply with [RLN](https://lip.logos.co/messaging/standards/core/17/rln-relay.html) constraints when sending segmented messages.

Reviewer: Not to address in this PR, but I'd like us to disuss this.

I recently thought: why is RLN compliance part of Reliable Channel API, and not Messaging API? 

If a network has RLN enabled, then one can't publish messages on it without RLN membership, and can't verify proofs without access to smart contract. In such case, Kernel API (relay/lighpush/filter/disc) still make sense, but Messaging API becomes useless.

So I wonder if it would make sense to implement it in Messaging API instead. Or am I not seeing something here?

Reviewer: Hmm, perhaps segmentation is the reason for introducing `Rate Limit Manager`. Without segmenting messages, we'll just get an error if attempt to send over the limit with with Messaging API.

On the other hand, `Rate Limit Manager`, should be not per-channel, but per-node 🤔

Author: I think_segmentation_ is something that we can leave for the latest phases, because segmentation is just doing the same but allowing larger messages.

Reviewer: > I think_segmentation_ is something that we can leave for the latest phases, because segmentation is just doing the same but allowing larger messages.

I'm not sure I'm following your reply 🤔

Author: We can implement the reliable channel API without segmentation and without RLN, in a very initial phase.
Later, we can add both segmentation and RLN, in further iterations.
I think this is reflected in [this](https://github.com/logos-messaging/pm/issues/412) issue's description.
Let me know if is clearer now :)

Author: As you suggested in our 1:1, let's have as is for now but continue the conversation in the following: https://github.com/logos-co/logos-lips/issues/321

┌────────────────────────────────────────────────────────────┐
│                      Logos Chat                            │
│              (Identity + Encryption + UX)                  │
└───────────────────────────┬────────────────────────────────┘
                            │
┌───────────────────────────▼────────────────────────────────┐
│                 Reliable Channel API                       │
│  ┌──────────────┐ ┌─────┐ ┌───────────────┐ ┌──────────┐   │
│  │ Segmentation │ │ SDS │ │ Rate Limit Mgr│ │Encryption│   │
│  │              │ │     │ │               │ │   Hook   │   │
│  └──────────────┘ └─────┘ └───────────────┘ └──────────┘   │
└───────────────────────────┬────────────────────────────────┘
                            │
┌───────────────────────────▼────────────────────────────────┐
│                    Messaging API                           │
│      (P2P Reliability, Relay, Filter, Lightpush, Store)    │
└────────────────────────────────────────────────────────────┘
```

## The Reliable Channel API

### Common

#### Common type definitions

This API considers the types defined by [MESSAGING-API](/standards/application/messaging-api.md) plus the following:

> ```yaml

Reviewer: It's better to define the implementation top-down. Start with `ReliableChannel`, then how to create a channel, and only then all types, configs, etc.

Current order is confusing. The reader is still trying to understand what's this API about, but has to read about some `MessageChunk` first.

Reviewer: I suggest this order:

1. Types definition 
    - `ReliableChannel`
    - its configs
2. Channels lifecycle
    - `createReliableChannel`
3. Channels usage
    - `channel.send(...)`
    - channel events

Author: Addressed in b03f1b8f247dea36c3e33e04c8c86c93588e4c45

Reviewer: @Ivansete-status changes in b03f1b8 don't look like what I proposed.
Do you disagree with something?

Reviewer: The problem I have with current order is that the first thing I see is `ReliableSendId`, then some `IEncryption` and `IPersistence`. As a reader, at this point these types make no sense to me, I am not understanding what they're for.

That's why I think it should be described TOP-DOWN. Start with things that a client will face first, e.g.:
1. high-level primitives like `ReliableChannel`
2. how to create a channel, i.e. `createReliableChannel`
3. how to send/receive messages on a reliable channel, i.e. methods of `ReliableChannel`

It should be kind of a logical story, wdyt?

Author: ah sorry, yes I agree 100%, I didn't execute the request properly. This should address it: https://github.com/logos-messaging/specs/commit/689bc0da1033cd191f7e8b7138aafae36bf4d7a7

types:

>   MessageChunk:

Reviewer: `MessageChunk` seem to be not referred to at all. Isn't it internal detail of `SEGMENTATION` protocol?

Author: Yes, absolutely. I completely overlooked that `SEGMENTATION` protocol and makes sense 100% to be a separated one. Addressed in https://github.com/logos-messaging/specs/commit/d85c71b7d12705e560a812ac86b187746b42e8c2

      lastChunkIndex:
        type: uint
        description: "All chunks of same message have the same value, i.e., num-chunks - 1."
      requestId:
        type: RequestId
        description: "The request id this chunk belongs to. This is generated by the send function and its type is defined in messaging-api."

  SyncStatusDetail:
    type: object
    description: "Provides a snapshot of the channel's synchronisation state."
    fields:
      received:
        type: uint
        description: "Number of messages successfully received."
      missing:
        type: uint
        description: "Number of messages detected as missing but not yet retrieved."
      lost:
        type: uint
        description: "Number of messages that could not be retrieved after all attempts."

  SyncStatus:
    type: enum
    description: "Indicates whether the channel has detected any outstanding missing messages."
    values:
      - synced: "No known missing messages (some may be permanently lost)."
      - syncing: "Actively attempting to retrieve one or more missing messages."

>   IEncryption:

Reviewer: I don't remember seeing the `I` notation anywhere else in specs, but correct me if I'm wrong.
```suggestion
  Encryption:
```

Author: The `I` is to emphasize that the implementation should be provided by the caller app. We can remove it if that causes confusion.

Reviewer: Yeah I know what `I` is about 😄 
I think it is an unnecessary fog for specs. Also, it's common in C#, but not in the languages we're using (Nim).

Author: Addressed in 6371ee653168205b5729376f9b09765f8eddad16

    type: object
    description: "Interface for a pluggable encryption mechanism.
    When provided, see ReliableChannelConfig, the API consumer MUST implement both encrypt and decrypt operations.
    Implementations MAY use different signatures than those described below, as long as each operation accepts a byte array and returns a byte array."
    fields:
      encrypt:
        type: function
        description: "Encrypts a byte payload. Returns the encrypted payload."
        parameters:

>           - name: plaintext

Reviewer: it's not "text" is it?
```suggestion
          - name: content
```

            type: array<byte>
        returns:
          type: result<array<byte>, error>
      decrypt:
        type: function
        description: "Decrypts a byte payload. Returns the decrypted payload."
        parameters:

>           - name: ciphertext

Reviewer: ```suggestion
          - name: payload
```

            type: array<byte>
        returns:
          type: result<array<byte>, error>
```

### Channel

#### Channel type definitions

```yaml
types:

>   ReliableEnvelope:

Reviewer: Does the spec care about this implementation detail type?

Author: Good catch! Addressed in eed8689530a5fcb5312c5b08faba0b89f44f741e

Reviewer: So `ReliableEnvelope` is just a message+channelId? If so, why can't it be just 2 separate arguments?

Author: I removed the `ReliableEnvelope` type. Now passing `arrat<byte>`. See b0967ac293bde44d94038278f84aaf55b3aaff7d and 73e0e0ba0ab47d7496227ccdbe2f59925f4e1334

  ReliableChannel:
    type: object
    description: "An entity that guarantees message delivery among all participants on a single content topic."
    fields:
      channelId:
        type: string
        description: "Unique identifier"
      messageEvents:
        type: ReliableMessageEvents
        description: "Event emitter for message-related events on this channel"
      syncStatus:
        type: SyncStatus
        description: "Current synchronisation status of the channel. Transitions between 'syncing' and 'synced' as missing messages are detected and resolved."

  ReliableChannelConfig:
    type: object
    fields:
      segmentationConfig:
        type: SegmentationConfig
        default: DefaultSegmentationConfig
        description: "Configuration for message segmentation."
      sdsConfig:
        type: SdsConfig
        default: DefaultSdsConfig
        description: "Configuration for Scalable Data Sync."
      rateLimitConfig:
        type: RateLimitConfig
        default: DefaultRateLimitConfig
        description: "Configuration for rate limit management."
      encryption:
        type: optional<IEncryption>
        default: none

>         description: "Optional pluggable encryption implementation. If none, messages are sent unencrypted."

Reviewer: I'd rather keep `encryption` out of "config". 
Config should be completely optional, while `encryption` will be provided most of the times.

Author: Ah yes I see.
We can have a separate _encryption setter_ function.
I would leave it as is for now and we can revisit the spec in the future once we have a proper implementation, if that sounds reasonable.

> 

Reviewer: Some of config parameters should be defined not per-channel, but per-node, e.g.:
- persistence provider (`historyBackend`)
- RLN configuration
- (maybe) segmentation `chunkSizeBytes`, because of certain network limits

Reviewer: Actually, all 3 `segmentationConfig`, `sdsConfig` and `rateLimitConfig` should be node-level configuration parameters. These are implementation details, which should not be exposed at Reliable Channel API.

Instead, this config could contain actual per-channel parameters, like `enable_store_queries` or smth. We could leave this empty for now, until we know what exactly we want to modify per-channel.

Author: Addressed in 77b933377ecb53ce0fc8c35cd29385cd74a9b0f7
Not adding `enable_store_queries` attribute for now and we'll get back to the spec as we develop.

  SegmentationConfig:
    type: object
    fields:
      chunkSizeBytes:
        type: uint
        default: 102400  # 100 KiB
        description: "Maximum chunk size in bytes.
        Messages larger than this value are split before SDS processing."

  SdsConfig:
    type: object
    fields:

>       historyBackend:

Reviewer: ```suggestion
      persistence:
```

>         type: string

Reviewer: This should be not a string, but interface, which could be provided externally. E.g. Status will want to use their encrypted storage.
```suggestion
        type: Persistence
```

Author: Addressed in 033f17af96489bd7de2f3f57c9ec010d43cf3dee

        default: "memory"
        description: "Backend for persisting the SDS local history. Implementations MAY support custom backends (e.g., 'memory', 'sqlite')."
      acknowledgementTimeoutMs:
        type: uint
        default: 5000
        description: "Time in milliseconds to wait for acknowledgement before retransmitting."
      maxRetransmissions:
        type: uint
        default: 5
        description: "Maximum number of retransmission attempts before considering delivery failed."
      causalHistorySize:
        type: uint
        default: 2
        description: "Number of message IDs to consider in the causal history. With longer value, a stronger correctness is guaranteed but it requires higher bandwidth and memory."

  RateLimitConfig:
    type: object
    fields:
      enabled:
        type: bool
        default: true
        description: "Whether rate limiting is enforced. SHOULD be true when RLN is active."
      epochSizeMs:
        type: uint
        default: 600000  # 10 minutes

>         description: "The epoch size used by the RLN relay, in milliseconds. Only the first message chunk is rate limited."

Reviewer: > Only the first message chunk is rate limited

and other chunks are sent over the quota? 😄

Author: I promise that I thought it was a great idea LOL!
Fixed in d7836f4b32341d1e1063622035c8921d8f706363
because it is the proper way to also consider chunks in the rate-limit.

```

#### Channel function definitions

```yaml
functions:

>   openChannel:

Reviewer: I think "open" is not the best word here, we're not really opening anything. Creating a channel == "give user nice interface" + "start internal periodic routines".
```suggestion
  createReliableChannel:
```

    description: "Opens a reliable channel. Sets up the required SDS state, segmentation, and encryption."

>     parameters:

Reviewer: 1. I don't see some of the main parameters I expect here. 
    What's defined in https://github.com/logos-messaging/specs/pull/89 looks sufficient to me.
    ```suggestion
          - name: channelId
            type: string
            description: "An identifier for the channel. All participants MUST use the same id."
          - name: senderId
            type: string
            description: "An identifier for this sender. SHOULD be unique and persisted between sessions."
          - name: contentTopic
            type: string
            description: "The content topic to use for the channel."
          - name: encryption
            type: IEncryption
            description: "Optional encryption/decryption interface. If not provided, messages are sent without encryption (passthrough)."
          - name: options
            type: ReliableChannelConfig
            description: "Configuration options for the Reliable Channel."
    ```

2. `ReliableChannel` should have some ties to certain `LogosDeliveryNode`. 
    I envision this as `channel = node.createChannel(...)`. Perhaps in the spec we can describe this by adding a node argument:
     ```yaml
    - name: node
       type: LogosDeliveryNode
       description: "The Logos Delivery node instance to use for sending and receiving messages."
    ```

Author: 1. I think is only missing the `senderId`. Addressed in ce68befbef5860d28d907671acce0f3fb37e7779

2. This is implicit in the spec. For example, in `messaging-api.md` we have:
```nim
  send:
    description: "Send a message through the network."
    parameters:
      - name: message
        type: MessageEnvelope
        description: "Parameters for sending the message."
    returns:
      type: result<RequestId, error>
```

whereas, in the current `liblogosdelivery` implementation, we have the following (will get simpler soon:)
```nim
proc logosdelivery_send(
    ctx: ptr FFIContext[Waku],
    callback: FFICallBack,
    userData: pointer,
    messageJson: cstring,
) {.ffi.} =
```

Notice that `ctx` represents the logos-delivery-node instance

Reviewer: > I think is only missing the `senderId`

- If we don't add `channelId`, then what will we pass to SDS `wrapOutgoingMessage`? 🤔 
    https://github.com/logos-messaging/nim-sds/blob/13559198b06dfb7d3bf8018c5f2f1e9ca9e54999/sds.nim#L84-L90
- If we don't add `contentTopic`, how is it chosen then?

Author: We would have both the `channelId` and `contentTopic` inside the `ReliableEnvelope` object passed to the `send` function.
In the current `liblogosdelivery` implementation it accepts a JSON. We can have a more type-safe solution to accept a `ReliableEnvelope` object.

Reviewer: As agreed in DM, `channelId` and `contentTopic` are required arguments of this function. It's resolved now.

      - name: channelConfig
        type: ReliableChannelConfig
        description: "Configuration for the channel."
    returns:
      type: result<ReliableChannel, error>

  closeChannel:
    description: "Closes a reliable channel and releases all associated resources and internal state."
    parameters:
      - name: channel
        type: ReliableChannel
        description: "The channel to close."
    returns:
      type: result<void, error>
```

#### Channel predefined values

```yaml
values:

  DefaultSegmentationConfig:
    type: SegmentationConfig
    fields:
      chunkSizeBytes: 102400  # 100 KiB

  DefaultSdsConfig:
    type: SdsConfig
    fields:
      historyBackend: "memory"
      acknowledgementTimeoutMs: 5000
      maxRetransmissions: 5
      causalHistorySize: 2

  DefaultRateLimitConfig:
    type: RateLimitConfig
    fields:
      enabled: true
      epochSizeMs: 600000

> ```

Reviewer: This is a duplication, all defaults are already defined in each config.
I think it's safe to remove this with no need to specify `default: ` field sin `ReliableChannelConfig`

Author: Addressed in 9a67839c925b162a1389d44eb44e70260c48f33b

#### Channel extended definitions

**State management**:

Each `ReliableChannel` MUST maintain internal state for SDS, including:
- A committed message log recording sent messages and their acknowledgement status.
- An outgoing buffer of unacknowledged message chunks pending confirmation.
- An incoming buffer for partially received segmented messages awaiting full reassembly.

The state MUST be persisted according to the `historyBackend` specified in `SdsConfig`.
The default `memory` backend does not survive process restarts; implementors providing persistence SHOULD use a durable backend.

**Encryption**:

The `encryption` field in `ReliableChannelConfig` is intentionally optional.
The Reliable Channel API is agnostic to encryption mechanisms.

When an `IEncryption` implementation is provided, it MUST be applied as described in the [Messaging extended definitions](#messaging-extended-definitions).

### Messaging

#### Messaging type definitions

```yaml
types:

>   ReliableMessageReceivedEvent:

Reviewer: What's a "reliable message"? Just "message" make more sense to me

Author: Addressed in 2fc897e6555908a1b4f65816f25f533dd450bb6b

    type: object
    description: "Event emitted when a complete message has been received and reassembled."
    fields:
      requestId:
        type: RequestId

>         description: "Identifier of the `send` operation that initiated the message sending"

Reviewer: I don't see `RequestId` type defined. Is it a "message ID" I guess, maybe it would be a better name?

Author: This is defined in messaging-api.md
Addressed in 89148cae12ece3515373a313a5ce57d78f6bcd36

    fields:
      requestId:
        type: RequestId
        description: "The request ID associated with the sent message"

  ReliableMessageSendErrorEvent:
    type: object
    description: "Event emitted when a message send operation fails after exhausting retransmission attempts."
    fields:
      requestId:
        type: RequestId
        description: "The request ID associated with the failed message"
      error:
        type: string
        description: "Error message describing what went wrong"

  ReliableMessageAcknowledgedEvent:
    type: object
    description: "Event emitted when the message has been definitively acknowledged via causal history."
    fields:
      requestId:
        type: RequestId
        description: "The request ID associated with the acknowledged message"

  ReliableIrretrievableMessageEvent:
    type: object
    description: "Event emitted when a missing message could not be retrieved after all store-query attempts."
    fields:
      requestId:
        type: RequestId
        description: "The request ID of the unrecoverable message"

  ReliableSyncStatusEvent:
    type: object

>     description: "Event emitted when the channel's sync status transitions between 'synced' and 'syncing'."

Reviewer: This event is emitted only when status changes, but not when anything in `SyncStatusDetail` changes?

Also, I'm not sure how exactly `SyncStatusDetail` could be used and what's we're providing it for. I would remove it from the initial implementation and add later on demand.

Author: Yup I agree 100%. Addressed in fabf0e0eb4e62603071e0d9a92a1bf1dfea7bf7c

    fields:
      status:
        type: SyncStatus
        description: "The new sync status"
      detail:
        type: SyncStatusDetail
        description: "Counters describing the current sync state"

  ReliableMessageEvents:
    type: event_emitter
    description: "Event source for reliable message events on a channel"
    events:
      "reliable:message:received":
        type: ReliableMessageReceivedEvent
      "reliable:message:sent":
        type: ReliableMessageSentEvent
      "reliable:message:acknowledged":
        type: ReliableMessageAcknowledgedEvent

>         description: "Message definitively acknowledged via causal history."

Reviewer: descriptions here are duplication

Author: I think addressed in bdab8f7cab9b6a8375e75071c0d797d016d579d0

      "reliable:message:send-error":
        type: ReliableMessageSendErrorEvent
      "reliable:message:irretrievable":
        type: ReliableIrretrievableMessageEvent
        description: "Missing message could not be retrieved after all store-query attempts."
      "reliable:sync:status":
        type: ReliableSyncStatusEvent
        description: "Channel sync status changed between 'synced' and 'syncing'."
```

#### Messaging function definitions

```yaml
functions:
  send:
    description: "Send a message reliably through the channel. Applies segmentation, SDS, and encryption (if configured) before dispatching."
    parameters:
      - name: channel

>         type: optional<ReliableChannel>

Reviewer: I expect that `send` is a method of `ReliableChannel`, so it can't be optional.
If anything, Messaging API can be used directly in such case
```suggestion
        type: ReliableChannel
```

Author: I was envisioning to only have one single `send` for Reliable API.
In fact, the app consumer should only use Reliable API and never bypass it to directly try to use Messaging API, i.e., Messaging API is hidden to the app.
 
The `optional` has the purpose of allowing sending ephemeral messages without delivery guarantees.

4. **Dispatch**: Send each chunk via the underlying [MESSAGING-API](/standards/application/messaging-api.md).

**Incoming message processing order**:

When a chunk is received from the network, the implementation MUST process it in the following order:

1. **Decrypt**: If an `IEncryption` implementation is provided, decrypt the chunk.
2. **Apply SDS**: Deliver the chunk to the SDS layer, which emits acknowledgements and detects gaps.
3. **Reassemble**: Once all chunks for a message have been received, reassemble and emit a `reliable:message:received` event.

**Retransmission**:

The SDS layer MUST retransmit unacknowledged chunks after `acknowledgementTimeoutMs`,
up to `maxRetransmissions` times.
If a chunk is not acknowledged after all retransmission attempts, a `reliable:message:send-error` event MUST be emitted.

**Rate limiting**:

When `RateLimitConfig.enabled` is `true`, the implementation MUST space chunk transmissions
to comply with the RLN epoch constraints defined in `epochSizeMs`.
Chunks MUST NOT be sent at a rate that would violate the RLN message rate limit for the active epoch.

## Components

> ### Segmentation

Reviewer: We should refer to https://github.com/logos-messaging/specs/pull/91.
I'll assign it to myself and get merged.

Author: Addressed in d85c71b7d12705e560a812ac86b187746b42e8c2
Please invite me to the _segmentation spec_ party too :)
